lib/pronto/rubocop/patch_cop.rb

```diff
@@ -14,10 +14,15 @@ def messages
         return [] unless valid?
 
         offenses.flat_map do |offense|
-          patch
+          offending_line = patch
             .added_lines
-            .select { |line| line.new_lineno == offense.line }
-            .map { |line| OffenseLine.new(self, offense, line).message }
+            .detect { |line| offense_includes?(offense, line.new_lineno) }
```

Comment:
I'm not 100% sure on using `detect` vs `select` -- normally, there's only going to be only one. But earlier, you could have multiple offenses reported -- what do you think about making it `select { }.uniq` instead?

Reply:
This method-chain is inside the call to `offenses.flat_map` - it should be getting executed once per offense, and producing a message for an `OffenseLine` that matches both the offense and the patch (if one exists).

If we updated the `detect { |line| offense_includes(offense, line.new_lineno) }` to be `select { |line| offense_includes?(offense, line.new_lineno) }.uniq`, we'd be calling `uniq` on an array of Line objects, which will all be different.. But if you intended that we unique the _messages_ afterward, I'm not sure if that would work correctly or not - it would certainly couple this logic to the definitions of those message strings though, and potentially do a lot of extra work constructing them all. Picture a spanning warning from lines 3-5003 of a large class that has been re-indented; we'd be constructing an OffenseLine for each of 5000 lines, constructing their messages, and `uniq`ing them back down to one afterward.

This would probably be clearer if the method-chain here were extracted to a private method like `first_relevant_line(patch, offense)` - I'd be happy to extract it if you like.

```
def messages
  return [] unless valid?
  
  offenses.flat_map do |offense|
    offending_line = first_relevant_line(patch, offense)
    if offending_line
      [OffenseLine.new(self, offense, offending_line).message]
    else
      []
    end
  end
end

private

# Find the first line of the patch that includes any line of the offense in question (nil if none do)
def first_relevant_line(patch, offense)
  patch
    .added_lines
    .detect { |line| offense_includes?(offense, line.new_lineno) }
end
```

I think the `flat_map` might also be confusing the issue - it shouldn't be necessary here, since every offense should only be presented once:

```
def messages
  return [] unless valid?
  
  offenses
    .map { |offense| first_relevant_message(patch, offense) }
    .compact
end

private

def first_relevant_message(patch, offense)
  offending_line = first_relevant_line(patch, offense)
  return nil unless offending_line

  OffenseLine.new(self, offense, offending_line).message
end
```

That's actually a _lot_ clearer, I'll go ahead and update.

Reply:
Thanks for the detailed explanation!